File: src/main/java/org/micoli/php/configuration/documentation/JavaDocumentationGenerator.java
package org.micoli.php.configuration.documentation;

import com.github.javaparser.JavaParser;
import com.github.javaparser.ParseResult;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
import com.github.javaparser.ast.body.MethodDeclaration;
import com.github.javaparser.javadoc.Javadoc;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Stream;

public class JavaDocumentationGenerator {

    public static class ClassInfo {
        private String name;
        private String description;
        private List<MethodInfo> methods = new ArrayList<>();

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }

        public List<MethodInfo> getMethods() {
            return methods;
        }
    }

    public static class MethodInfo {
        private String name;
        private String signature;
        private String description;
        private List<ParameterInfo> parameters = new ArrayList<>();
        private String returnDescription;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getSignature() {
            return signature;
        }

        public void setSignature(String signature) {
            this.signature = signature;
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }

        public List<ParameterInfo> getParameters() {
            return parameters;
        }

        public String getReturnDescription() {
            return returnDescription;
        }

        public void setReturnDescription(String returnDescription) {
            this.returnDescription = returnDescription;
        }
    }

    public static class ParameterInfo {
        private String name;
        private String description;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }
    }

    public static Map<String, JavaDocumentationGenerator.ClassInfo> extractJavadocs(String sourcePath) {
        Map<String, JavaDocumentationGenerator.ClassInfo> classInfos = new HashMap<>();

        try (Stream<Path> paths = Files.walk(Paths.get(sourcePath))) {
            List<Path> javaFiles = paths.filter(Files::isRegularFile)
                    .filter(path -> path.toString().endsWith(".java"))
                    .toList();

            for (Path javaFile : javaFiles) {
                try (FileInputStream in = new FileInputStream(javaFile.toFile())) {
                    ParseResult<CompilationUnit> parseResult = new JavaParser().parse(in);
                    if (parseResult.getResult().isEmpty()) {
                        continue;
                    }
                    CompilationUnit cu = parseResult.getResult().get();
                    for (ClassOrInterfaceDeclaration classDecl : cu.findAll(ClassOrInterfaceDeclaration.class)) {
                        if (classDecl.getNameAsString().contains("Error")
                                || classDecl.getNameAsString().contains("Exception")) {
                            continue;
                        }
                        JavaDocumentationGenerator.ClassInfo classInfo = new JavaDocumentationGenerator.ClassInfo();
                        classInfo.setName(classDecl.getNameAsString());

                        Optional<Javadoc> classJavadoc = classDecl.getJavadoc();
                        classJavadoc.ifPresent(javadoc -> classInfo.setDescription(
                                javadoc.getDescription().toText()));

                        for (MethodDeclaration methodDecl : classDecl.getMethods()) {
                            if (methodDecl.isPublic()) {
                                JavaDocumentationGenerator.MethodInfo methodInfo =
                                        new JavaDocumentationGenerator.MethodInfo();
                                methodInfo.setName(methodDecl.getNameAsString());
                                methodInfo.setSignature(methodDecl.getDeclarationAsString(false, false, true));

                                Optional<com.github.javaparser.javadoc.Javadoc> methodJavadoc = methodDecl.getJavadoc();
                                if (methodJavadoc.isPresent()) {
                                    methodInfo.setDescription(
                                            methodJavadoc.get().getDescription().toText());

                                    methodJavadoc.get().getBlockTags().stream()
                                            .filter(tag -> tag.getType()
                                                    == com.github.javaparser.javadoc.JavadocBlockTag.Type.PARAM)
                                            .forEach(tag -> {
                                                JavaDocumentationGenerator.ParameterInfo paramInfo =
                                                        new JavaDocumentationGenerator.ParameterInfo();
                                                paramInfo.setName(tag.getName().orElse(""));
                                                paramInfo.setDescription(
                                                        tag.getContent().toText());
                                                methodInfo.getParameters().add(paramInfo);
                                            });

                                    methodJavadoc.get().getBlockTags().stream()
                                            .filter(tag -> tag.getType()
                                                    == com.github.javaparser.javadoc.JavadocBlockTag.Type.RETURN)
                                            .findFirst()
                                            .ifPresent(tag -> methodInfo.setReturnDescription(
                                                    tag.getContent().toText()));
                                }

                                classInfo.getMethods().add(methodInfo);
                            }
                        }

                        classInfos.put(classInfo.getName(), classInfo);
                    }
                } catch (Exception e) {
                    throw new RuntimeException(e.getMessage());
                }
            }
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage());
        }

        return classInfos;
    }

    public static String generateMarkdownDocumentation(String sourcePath) {
        Map<String, ClassInfo> classInfos = JavaDocumentationGenerator.extractJavadocs(sourcePath);
        StringBuilder markdown = new StringBuilder();
        for (ClassInfo classInfo : classInfos.values()) {
            markdown.append("#### ")
                    .append("`")
                    .append(classInfo.getName())
                    .append("`")
                    .append("\n\n");
            if (classInfo.getDescription() != null) {
                markdown.append(classInfo.getDescription()).append("\n\n");
            }

            if (!classInfo.getMethods().isEmpty()) {
                for (MethodInfo methodInfo : classInfo.getMethods()) {
                    markdown.append("- `").append(methodInfo.getSignature()).append("`\n");
                    if (methodInfo.getDescription() != null) {
                        markdown.append("  ")
                                .append(methodInfo.getDescription().replaceAll("\n", " "))
                                .append("\n");
                    }

                    if (!methodInfo.getParameters().isEmpty()) {
                        for (ParameterInfo paramInfo : methodInfo.getParameters()) {
                            markdown.append("   - `")
                                    .append(paramInfo.getName())
                                    .append("`: ")
                                    .append(paramInfo.getDescription().replaceAll("\n", " "))
                                    .append("\n");
                        }
                        markdown.append("\n");
                    }

                    if (methodInfo.getReturnDescription() != null) {
                        markdown.append(methodInfo.getReturnDescription()).append("\n\n");
                    }
                }
            }
        }

        return markdown.toString();
    }
}

File: src/main/kotlin/org/micoli/php/configuration/documentation/JavaDocumentationGenerator.kt
package org.micoli.php.configuration.documentation

import com.github.javaparser.JavaParser
import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration
import com.github.javaparser.javadoc.Javadoc
import com.github.javaparser.javadoc.JavadocBlockTag
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.function.Consumer

object JavaDocumentationGenerator {
    fun extractJavadocs(sourcePath: String): MutableMap<String?, ClassInfo>? {
        val classInfos: MutableMap<String?, ClassInfo> = HashMap()

        try {
            Files.walk(Paths.get(sourcePath)).use { paths ->
                val javaFiles =
                    paths
                        .filter { path: Path? -> path != null }
                        .filter { path: Path -> Files.isRegularFile(path) }
                        .filter { path: Path -> path.toString().endsWith(".java") }
                        .toList()
                for (javaFile in javaFiles) {
                    try {
                        FileInputStream(javaFile.toFile()).use { `in` ->
                            val parseResult = JavaParser().parse(`in`)
                            if (parseResult.getResult().isEmpty) {
                                return null
                            }
                            val cu = parseResult.getResult().get()
                            for (classDecl in cu.findAll(ClassOrInterfaceDeclaration::class.java)) {
                                if (classDecl.nameAsString.contains("Error") ||
                                    classDecl.nameAsString.contains("Exception")) {
                                    continue
                                }
                                val classInfo = ClassInfo()
                                classInfo.name = classDecl.nameAsString

                                val classJavadoc = classDecl.javadoc
                                classJavadoc.ifPresent(
                                    Consumer { javadoc: Javadoc? ->
                                        classInfo.description = javadoc!!.description.toText()
                                    })

                                for (methodDecl in classDecl.methods) {
                                    if (methodDecl.isPublic) {
                                        val methodInfo = MethodInfo()
                                        methodInfo.name = methodDecl.nameAsString
                                        methodInfo.signature =
                                            methodDecl.getDeclarationAsString(false, false, true)

                                        val methodJavadoc = methodDecl.javadoc
                                        if (methodJavadoc.isPresent) {
                                            methodInfo.description =
                                                methodJavadoc.get().description.toText()

                                            methodJavadoc
                                                .get()
                                                .blockTags
                                                .stream()
                                                .filter { tag: JavadocBlockTag? ->
                                                    (tag!!.type == JavadocBlockTag.Type.PARAM)
                                                }
                                                .forEach { tag: JavadocBlockTag? ->
                                                    val paramInfo = ParameterInfo()
                                                    paramInfo.name = tag!!.name.orElse("")
                                                    paramInfo.description = tag.content.toText()
                                                    methodInfo.parameters.add(paramInfo)
                                                }

                                            methodJavadoc
                                                .get()
                                                .blockTags
                                                .stream()
                                                .filter { tag: JavadocBlockTag? ->
                                                    (tag!!.type == JavadocBlockTag.Type.RETURN)
                                                }
                                                .findFirst()
                                                .ifPresent(
                                                    Consumer { tag: JavadocBlockTag? ->
                                                        methodInfo.returnDescription =
                                                            tag!!.content.toText()
                                                    })
                                        }

                                        classInfo.methods.add(methodInfo)
                                    }
                                }

                                classInfos[classInfo.name] = classInfo
                            }
                        }
                    } catch (e: Exception) {
                        throw RuntimeException(e.message)
                    }
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e.message)
        }

        return classInfos
    }

    fun generateMarkdownDocumentation(sourcePath: String): String {
        val classInfos = extractJavadocs(sourcePath) ?: return ""
        val markdown = StringBuilder()
        for (classInfo in classInfos.values) {
            markdown.append("#### ").append("`").append(classInfo.name).append("`").append("\n\n")
            if (classInfo.description != null) {
                markdown.append(classInfo.description).append("\n\n")
            }

            if (!classInfo.methods.isEmpty()) {
                for (methodInfo in classInfo.methods) {
                    markdown.append("- `").append(methodInfo.signature).append("`\n")
                    if (methodInfo.description != null) {
                        markdown
                            .append("  ")
                            .append(methodInfo.description!!.replace("\n".toRegex(), " "))
                            .append("\n")
                    }

                    if (!methodInfo.parameters.isEmpty()) {
                        for (paramInfo in methodInfo.parameters) {
                            markdown
                                .append("   - `")
                                .append(paramInfo.name)
                                .append("`: ")
                                .append(paramInfo.description!!.replace("\n".toRegex(), " "))
                                .append("\n")
                        }
                        markdown.append("\n")
                    }

                    if (methodInfo.returnDescription != null) {
                        markdown.append(methodInfo.returnDescription).append("\n\n")
                    }
                }
            }
        }

        return markdown.toString()
    }

    class ClassInfo {
        var name: String? = null
        var description: String? = null
        val methods: MutableList<MethodInfo> = ArrayList()
    }

    class MethodInfo {
        var name: String? = null
        var signature: String? = null
        var description: String? = null
        val parameters: MutableList<ParameterInfo> = ArrayList()
        var returnDescription: String? = null
    }

    class ParameterInfo {
        var name: String? = null
        var description: String? = null
    }
}
